### PR TITLE
Parse&resolve&complete rust paths in rustdoc links

### LINKED ISSUE #3591: Allow navigation in rustdocs of the standard library
Rustdocs from the standard library contain a lot of links to the other places of the documentation:

image

It would be great to be able to navigate to those via Cmd + click (Ctrl + click) similar to the way it's possible to navigate to the method definition currently.

### LINKED ISSUE #4518: Autocomplete cargo doc paths/links to types in doc comment
If you write something like the following in a doc comment

```rust
/// An error returned by [`RefCell::try_borrow`](struct.RefCell.html#method.try_borrow).
#[stable(feature = "try_borrow", since = "1.13.0")]
pub struct BorrowError {
    _private: (),
}
```

The `struct.RefCell.html#method.try_borrow` your documentation site will link to the `try_borrow` method when you click on the `RefCell::try_borrow` link.

Since the naming convention for pages / links within a cargo doc generated site is consistent - could intellij-rust support autocomplete for these links?

This would make it much easier to write documentation that references other types/methods within your library.

Thanks!

### PR DESCRIPTION
**Pull request by vlad20012**
Fixes #3591, fixes #4518

https://user-images.githubusercontent.com/3221931/122210853-c46c5f00-ceae-11eb-961a-a05f98d529cf.mp4

changelog: Support [intra-doc links](https://doc.rust-lang.org/rustdoc/write-documentation/linking-to-items-by-name.html)

### CODE REVIEW

**Comment by Kobzol** [comment]
I *think* that this closes https://github.com/intellij-rust/intellij-rust/issues/3591 and https://github.com/intellij-rust/intellij-rust/issues/4518?

File: src/main/kotlin/org/rust/lang/doc/psi/RsDocLinkDestinationParser.kt [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+import org.rust.lang.core.lexer.RsLexer
+import org.rust.lang.core.parser.RustParser
+import org.rust.lang.doc.psi.RsDocElementTypes.DOC_LINK_DEFINITION
+
+/**
+ * `fn@mod1::mod2::func#anchor`
+ *  ~~~ prefix         ~~~~~~~ suffix
+ *     ~~~~~~~~~~~~~~~~ text
+ */
+private class LinkInfo private constructor(
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/doc/psi/RsDocLinkDestinationParser.kt:26]: Maybe `LinkTextParts`? :thinking:

File: src/main/kotlin/org/rust/lang/doc/psi/RsDocLinkDestinationParser.kt [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+            .trimBackticks()
+            ?.removeHashAnchor()
+            ?.removeDisambiguator()
+            ?.takeIf { canBeCorrectLink(it.text) }
+    }
+
+    // "`func`" -> "func"
+    // Remove any number of backticks at the beginning and end.
+    // Rustdoc also removes backticks in the middle, but we don't support it
+    private fun LinkInfo.trimBackticks(): LinkInfo? {
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/doc/psi/RsDocLinkDestinationParser.kt:71]: I'd prefer moving these extension on top-level near `LinkInfo` or into `LinkInfo` itself to make more explicit that they don't use any context outside of `LinkInfo`

Comment (dima74) [reply src/main/kotlin/org/rust/lang/doc/psi/RsDocLinkDestinationParser.kt:75]: Do you mean only `trimBackticks` or also `removeHashAnchor` etc? Actually `RsDocLinkDestinationParser` is `object`, so there can't be any context anyway

Comment (vlad20012) [reply src/main/kotlin/org/rust/lang/doc/psi/RsDocLinkDestinationParser.kt]: Nevermind. I thought a little better and now it seems to me that the way it is even better.

File: src/test/resources/org/rust/lang/core/parser/fixtures/complete/doc_comments.rs [hunk 3]
```diff
@@ -87,11 +87,28 @@
...
+/// [rust path with disambiguator 4]: foo::bar::baz!
+/// [rust path with hash]: foo::bar::baz#quux
+/// [rust path with disambiguator and hash 1]: foo::bar::baz!()#hash
+/// [rust path with disambiguator and hash 2]: fn@foo::bar::baz#hash
+/// [rust path generics]: Vec<i32>
+/// [rust path generics]: Vec<i32>::new()
+/// [rust path in backticks 1]: foo`
+/// [rust path in backticks 2]: `foo
+/// [rust path in backticks 3]: `foo`
+/// [rust path in backticks 4]: `fn@Vec::new#hash`
```
Comment (vlad20012) [inline src/test/resources/org/rust/lang/core/parser/fixtures/complete/doc_comments.rs:111]: Let's add more tests with more weird backticks

**Comment by dima74** [comment]
bors r=vlad20012

**Pull request merged**